Fixes the removal of the DefaultValue attribute from ListBox.ItemHeight property

dkazennov opened the pull request:
Fixes #4463 


## Proposed changes

- Changed the value of the DefaultListBoxItemHeight property. It is affected by the DPI.
- Therefore finishes #8518 

## Customer Impact

- Fixes the Issue which impacts user experience, breaks DPI and Font scaling. For the newly created control at 100% scaling, designer should generate the same value as is defined by the Default. There is no extra code.


## Regression? 

- No

## Risk

- Minimal

<!-- end TELL-MODE -->


## Screenshots <!-- Remove this section if PR does not change UI -->

### Before

![image](https://user-images.githubusercontent.com/15823268/104525385-e63a8080-55b4-11eb-8018-1cd886b658a0.png)

### After

![image](https://user-images.githubusercontent.com/102961955/223371358-5d3f24a5-49c1-4554-b21e-77b1382560d7.png)


## Test methodology <!-- How did you ensure quality? -->

- Manual (Issue can be tested via substitution of dlls, complete restart of the VS recommended)
- Unit test

## Test environment(s) <!-- Remove any that don't apply -->

- dotnet 8.0.100-alpha.1.22607.6


###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/dotnet/winforms/pull/8755)

dkazennov commented on src/System.Windows.Forms/src/System/Windows/Forms/ListBox.cs:
```diff
@@ -2225,7 +2225,13 @@ public void SetSelected(int index, bool value)
+                return ItemHeight != DefaultListBoxItemHeight;
+            }
+
+            return ItemHeight != GetItemHeight(0);
```
> When setting new height with `DrawMode.Normal`, does it get serialized correctly? For example, if I set new height as 20, I assume both `ItemHeight` and `GetItemHeight(0)` would return 20, and it won't be serialized. Or is it somehow works?

When we get ItemHeight value, we either get _itemHeight directly or get GetItemHeight() return value, which in turn gets height via request PInvoke.SendMessage (or return _itemHeight if handle is not created).  

Oh, while I was writing the reply I saw that we should rather have `GetItemHeight(0) !=DefaultListBoxItemHeight` for `DrawMode.Normal`.

dkazennov replied:
@dmitrii-drobotov you are correct. This is wrong and it affect tests in a right way. I will change it.
But more importantly I double checked the problem and I don't have reproduction with #8518 fix (on main branch now). I'll ask additional information about conditions under that the issue was reproduced.

dkazennov commented:
We have found that this fix should consider the DPI value.

dkazennov commented:
I've checked this fix with 100%, 125% and 150% display resolution scale. There is no extra code in any mode.

dkazennov commented on src/System.Windows.Forms/src/System/Windows/Forms/ListBox.cs:
```diff
@@ -396,7 +396,7 @@ private static int DefaultListBoxItemHeight
                 if (s_defaultListBoxItemHeight == -1)
                 {
-                    s_defaultListBoxItemHeight = DefaultFont.Height;
+                    s_defaultListBoxItemHeight = DefaultFont.Height - 1;
```
> Why is `-1` needed?

I expected that we need a small adjustments for the default value (keep it relative but adjust it to be in line with DPI settings).  So I have found out that with the default value being `equal DefaultFont.Height - 1` this fix works in all DPI modes, and all tests passes.

Tanya-Solyanik replied:
I expected that this value can be calculated based on the `DrawMode`. If the mode is not OwnerDraw, then it would be read from the native control.

Tanya-Solyanik commented on src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/ListBoxTests.cs:
```diff
@@ -1836,12 +1838,12 @@ public void ListBox_ItemHeight_ShouldSerializeValue_Success()
-            Assert.True(property.ShouldSerializeValue(control));
+            control.ItemHeight = 17;
+            Assert.Equal(17, control.ItemHeight);
+            //Assert.True(property.ShouldSerializeValue(control));
```
Should this line be enabled?

dkazennov replied:
> Should this line be enabled?

Yes, indeed. Fixed it.

dmitrii-drobotov commented on src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/ListBoxTests.cs:
```diff
@@ -1836,12 +1838,12 @@ public void ListBox_ItemHeight_ShouldSerializeValue_Success()
             property.ResetValue(control);
-            Assert.Equal(Control.DefaultFont.Height, control.ItemHeight);
+            Assert.Equal(s_defaultListBoxItemHeight, control.ItemHeight);
             Assert.False(property.ShouldSerializeValue(control));
```
Consider adding more cases to ShouldSerialize and Reset unit tests - check with all DrawMode values, set a different default font, perhaps add `control.CreateControl()` so that handle gets created.

dkazennov commented:
@dmitrii-drobotov I have changed existing tests and added the new test `ListBox_ItemHeight_ResetValue_Success_If_DefaultListBoxItemHeight_IsChanged` which shows that with different application default fonts and different draw modes we should have `ListBox.DefaultListBoxItemHeight` to be equal `Control.DefaultFont.Height - 1`.

v-elnovikova commented on src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/ListBoxTests.cs:
```diff
@@ -2050,13 +2111,13 @@ public void ListBox_Padding_SetWithHandler_CallsPaddingChanged()
-            yield return new object[] { DrawMode.OwnerDrawFixed, BorderStyle.FixedSingle, Control.DefaultFont.Height + extra };
-            yield return new object[] { DrawMode.OwnerDrawFixed, BorderStyle.None, Control.DefaultFont.Height };
+            yield return new object[] { DrawMode.OwnerDrawFixed, BorderStyle.Fixed3D, s_defaultListBoxItemHeight+ extra };
+            yield return new object[] { DrawMode.OwnerDrawFixed, BorderStyle.FixedSingle, s_defaultListBoxItemHeight+ extra };
```
Small nit: missed space before binary operator here, and line 2115 too.

v-elnovikova reviewed (approved):
Looks good!

dkazennov commented:
Requested CTI testing just to be sure.

dreddy-work commented:
@dkazennov, please rebase this PR.

dkazennov commented:
> @dkazennov, please rebase this PR.

Done it. Resolved all conflicts.

dreddy-work commented:
I am not inclined to go with this solution. Will sync with @Tanya-Solyanik and figure out what we should be doing here. Root cause must be a result of our default Font change.

lonitra commented:
Closing as the approach to fix this issue seems to be superseded by https://github.com/dotnet/winforms/pull/10142
